File: 759-employee-free-time/759-employee-free-time.java
/*
// Definition for an Interval.
class Interval {
    public int start;
    public int end;

    public Interval() {}

    public Interval(int _start, int _end) {
        start = _start;
        end = _end;
    }
};
*/

class Solution {
    public List<Interval> employeeFreeTime(List<List<Interval>> schedule) {
        List<Interval> res = new ArrayList<>();
        PriorityQueue<Interval> pq = new PriorityQueue<>((a, b) -> a.start == b.start ? b.end - a.end : a.start - b.start);
        for(List<Interval> p : schedule) {
            for(Interval i : p) {
                pq.offer(i);
            }
        }
        Interval curr = pq.poll();
        while(!pq.isEmpty()) {
            if(curr.end >=  pq.peek().start) {
                curr.end = Math.max(curr.end, pq.poll().end);
            } else {
                res.add(new Interval(curr.end, pq.peek().start));
                curr = pq.poll();
            }
        }
       
        return res;
    }
}

File: 759-employee-free-time/759-employee-free-time.py
"""
# Definition for an Interval.
class Interval:
    def __init__(self, start: int = None, end: int = None):
        self.start = start
        self.end = end
"""

class Solution:
    def employeeFreeTime(self, schedule: '[[Interval]]') -> '[Interval]':
        lst = sorted([i for s in schedule for i in s], key=lambda x : x.start)
        prev, res = lst[0], []
        for i in lst[1:]:
            if prev.end >= i.start:
                prev.end = max(prev.end, i.end)
            else:
                res.append(Interval(prev.end, i.start))
                prev = i
        return res